### ISSUE TITLE
react/no-unused-prop-types false positive when using memo

### ISSUE THREAD

**Issue by markholland** [jsx-eslint/eslint-plugin-react#3403]
Here's an [example using CRA](https://github.com/markholland/eslint-plugin-react-example) 

**Expected**: No warning

**Actual**:
- `npm run lint`
```
eslint-example/src/App.js
  23:3  warning  'showFirst' PropType is defined but prop is never used  react/no-unused-prop-types
```

This started when going from version `7.30.0` to `7.30.1`.

- If I remove `memo` the warning goes away
- If I wrap the ternary in the return with a fragment the warning goes away but then I'd have to silence `react/jsx-no-useless-fragment` so not helpful either.

**Comment by ljharb** [comment]
Looking at https://github.com/markholland/eslint-plugin-react-example/blob/main/src/App.js, both of those render functions should either be components defined at module level, or inline jsx. With one of those changes, what happens?

**Comment by markholland** [comment]
For us this is a list row component so splitting them at module level wouldn't make sense, yes we could inline but each function has several possible jsx outcomes so made sense to split them into two.

We've rewritten our component to resolve this. Simply felt odd that an unrelated patch release surfaced this warning.

Thanks for the help ❤️

**Comment by ljharb** [comment]
I agree, this does seem like a bug.

**Comment by godlanbo** [comment]
I also encountered this problem, is there any solution now, besides turning off the rules

**Comment by ljharb** [comment]
@godlanbo nope, if there was a solution then the issue would be closed.

**Comment by aynzad** [comment]
> I also encountered this problem, is there any solution now, besides turning off the rules

the only way for now is to use `memo` on export, something like this:

```
import React, { memo } from "react";
import PropTypes from "prop-types";

import "./App.css";

const App =({ showFirst }) => {
  const renderFirst = () => (
    <div>
      <p>First</p>
    </div>
  );

  const renderSecond = () => (
    <div>
      <p>Second</p>
    </div>
  );

  return showFirst ? renderFirst() : renderSecond();
};

App.propTypes = {
  showFirst: PropTypes.bool.isRequired,
};

export default memo(App);
```
